pkg/s3select/json/reader.go

```diff
@@ -17,186 +17,49 @@
-			}
-			output, err = toSingleLineJSON(input, jsonKey, value)
-			input = output
-			return err == nil
-		})
-	case result.IsArray():
-		i := 0
-		result.ForEach(func(key, value gjson.Result) bool {
-			if currentKey == "" {
-				panic("currentKey is empty")
-			}
-
-			indexKey := currentKey + "." + strconv.Itoa(i)
-			output, err = toSingleLineJSON(input, indexKey, value)
-			input = output
-			i++
-			return err == nil
-		})
-	default:
-		output, err = sjson.Set(input, currentKey, result.Value())
-	}
-
-	return output, err
-}
-
-type objectReader struct {
-	reader io.Reader
-	err    error
-
-	p     []byte
-	start int
-	end   int
-
-	escaped     bool
-	quoteOpened bool
-	curlyCount  uint64
-	endOfObject bool
-}
-
-func (or *objectReader) objectEndIndex(p []byte, length int) int {
-	for i := 0; i < length; i++ {
-		if p[i] == '\\' {
-			or.escaped = !or.escaped
-			continue
-		}
-
-		if p[i] == '"' && !or.escaped {
-			or.quoteOpened = !or.quoteOpened
-		}
-
-		or.escaped = false
-
-		switch p[i] {
-		case '{':
-			if !or.quoteOpened {
-				or.curlyCount++
-			}
-		case '}':
-			if or.quoteOpened || or.curlyCount == 0 {
-				break
-			}
-
-			if or.curlyCount--; or.curlyCount == 0 {
-				return i + 1
-			}
-		}
-	}
-
-	return -1
-}
-
-func (or *objectReader) Read(p []byte) (n int, err error) {
-	if or.endOfObject {
-		return 0, io.EOF
-	}
-
-	if or.p != nil {
-		n = copy(p, or.p[or.start:or.end])
-		or.start += n
-		if or.start == or.end {
-			// made full copy.
-			or.p = nil
-			or.start = 0
-			or.end = 0
-		}
-	} else {
-		if or.err != nil {
-			return 0, or.err
-		}
-
-		n, err = or.reader.Read(p)
-		or.err = err
-		switch err {
-		case nil:
-		case io.EOF, io.ErrUnexpectedEOF, io.ErrClosedPipe:
-			or.err = io.EOF
-		default:
-			return 0, err
-		}
-	}
-
-	index := or.objectEndIndex(p, n)
-	if index == -1 || index == n {
-		return n, nil
-	}
-
-	or.endOfObject = true
-	if or.p == nil {
-		or.p = p
-		or.start = index
-		or.end = n
-	} else {
-		or.start -= index
-	}
-
-	return index, nil
-}
-
-func (or *objectReader) Reset() error {
-	or.endOfObject = false
-
-	if or.p != nil {
-		return nil
-	}
-
-	return or.err
-}
-
 // Reader - JSON record reader for S3Select.
 type Reader struct {
 	args         *ReaderArgs
-	objectReader *objectReader
+	decoder      *jstream.Decoder
+	objectReader chan *jstream.MetaValue
 	readCloser   io.ReadCloser
 }
 
 // Read - reads single record.
 func (r *Reader) Read() (sql.Record, error) {
-	if err := r.objectReader.Reset(); err != nil {
-		return nil, err
+	v, ok := <-r.objectReader
+	if !ok {
+		if err := r.decoder.Err(); err != nil {
+			return nil, errJSONParsingError(err)
+		}
+		return nil, io.EOF
 	}
 
-	data, err := ioutil.ReadAll(r.objectReader)
-	if err != nil {
-		return nil, errJSONParsingError(err)
-	}
+	var data []byte
+	var err error
 
-	data = bytes.TrimSpace(data)
-	if len(data) == 0 {
-		return nil, io.EOF
+	if v.ValueType == jstream.Object {
+		data, err = json.Marshal(v.Value)
```

Comment:
This "re-marshalling" is a problem - it ends up re-ordering the keys from the input.

```shell
$ mc cat mys3/sqlselectapi/1.json
{ "Id" : 1, "name" : "q", "color" : "blue" }
{ "name" : "q", "color" : "blue" ,"Id" : 2}
{ "color" : "blue" ,"name" : "q",  "Id" : 3}
{ "Id" : 4 , "name" : "q", "color" : "blue", "Tele": "iphone" }
$ mc sql --query 'select * from s3object s' mys3/sqlselectapi/1.json
{"Id":1,"name":"q","color":"blue"}
{"name":"q","color":"blue","Id":2}
{"color":"blue","name":"q","Id":3}
{"Id":4,"name":"q","color":"blue","Tele":"iphone"}
$ mc sql --query 'select * from s3object s' myminio/sqlselectapi/1.json
{"Id":1,"color":"blue","name":"q"}
{"Id":2,"color":"blue","name":"q"}
{"Id":3,"color":"blue","name":"q"}
{"Id":4,"Tele":"iphone","color":"blue","name":"q"}
```
```

Reply:
So I discussed this @donatello since S3 also modifies the data, we don't need to comply with the order of keys. The output here is JSON and doesn't have a predefined order, to begin with. Since we cannot guarantee that order in the first place.

Reply:
So the output is expected and we will be fine.

Reply:
No, S3's implementation only removes whitespace, but keeps the input ordering. This distinction is important - S3 mentions this quite clearly in the documentation, quoting from [here](https://docs.aws.amazon.com/AmazonS3/latest/dev/s3-glacier-select-sql-reference-select.html):

> The first form with * (asterisk) returns every row that passed the WHERE clause, as-is.

If the user requests CSV output serialization, this behavior will cause the server to return garbage:

```
$ mc cat mys3/sqlselectapi/7.json
{"name":"q","Id":1,"color":"blue"}
{"name":"s","Id":2,"color":"blue"}
{"name":"t","Id":3,"color":"blue"}
$ mc sql --query 'select * from s3object s' mys3/sqlselectapi/7.json
{"name":"q","Id":1,"color":"blue"}
{"name":"s","Id":2,"color":"blue"}
{"name":"t","Id":3,"color":"blue"}
$ mc sql --query 'select * from s3object s' myminio/sqlselectapi/7.json
{"Id":1,"color":"blue","name":"q"}
{"Id":2,"color":"blue","name":"s"}
{"Id":3,"color":"blue","name":"t"}
```

Imagine the last two commands above requiring CSV output. It will be junk if the application assumes S3 behaviour.

Reply:
@donatello - why would it be junk? can you show me how it will be junk?

Reply:
I mean in CSV output, for the above file, S3 would give:

```
q,1,blue
s,2,blue
t,3,blue
```

We would give:

```
1,blue,q
2,blue,s
3,blue,t
```

Now the application that is expecting S3 behaviour is expecting to see Id (numeric) column as the second one, but minio returns it as the first one, and it does not work. 

The only "predictable" ordering is to keep it the same as the input (other than sorting of course, but that is significantly more expensive and not the route that S3 chose).